File: backend/registrations/printBadge.mjs
import { UpdateCommand, GetCommand } from "@aws-sdk/lib-dynamodb";
import { db, REGISTRATIONS_TABLE } from "./shared/db.mjs";
import { ok, badRequest, notFound, serverError, cors } from "./shared/response.mjs";
import { audit, callerFromEvent } from "./shared/auth.mjs";

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const caller = callerFromEvent(event);
  const { id } = event.pathParameters ?? {};
  if (!id) return badRequest("id is required");

  try {
    const existing = await db.send(new GetCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
    }));
    if (!existing.Item) return notFound("Registration not found");

    const now = new Date().toISOString();
    const newCount = (existing.Item.badgePrintCount ?? 0) + 1;

    await db.send(new UpdateCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
      UpdateExpression: "SET badgePrintedAt = :t, badgePrintCount = :c, updatedAt = :t",
      ExpressionAttributeValues: { ":t": now, ":c": newCount },
    }));
    await audit("badge.printed", {
      entity: "participant", entityId: id,
      actorId: caller.actorId, actorLabel: caller.actorLabel,
    });
    return ok({ badgePrintedAt: now, badgePrintCount: newCount });
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/registrations/printBadge.py
import logging

from shared.auth import audit, caller_from_event, iso_now
from shared.db import registrations_table
from shared.response import bad_request, cors, not_found, ok, server_error, with_cors

log = logging.getLogger()


@with_cors
def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    caller = caller_from_event(event)
    reg_id = (event.get("pathParameters") or {}).get("id")
    if not reg_id:
        return bad_request("id is required")

    try:
        existing = registrations_table().get_item(Key={"registrationId": reg_id}).get("Item")
        if not existing:
            return not_found("Registration not found")

        now = iso_now()
        new_count = int(existing.get("badgePrintCount") or 0) + 1

        registrations_table().update_item(
            Key={"registrationId": reg_id},
            UpdateExpression="SET badgePrintedAt = :t, badgePrintCount = :c, updatedAt = :t",
            ExpressionAttributeValues={":t": now, ":c": new_count},
        )
        audit("badge.printed", entity="participant", entity_id=reg_id,
              actor_id=caller["actorId"], actor_label=caller["actorLabel"])
        return ok({"badgePrintedAt": now, "badgePrintCount": new_count})
    except Exception:
        log.exception("printBadge failed")
        return server_error()
